### ISSUE TITLE
Tile edges visible

### ISSUE THREAD

**Issue by MindFreeze** [openseadragon/openseadragon#755]
I am using IIIF but don't think this matters. The tile edges are visible especially if I zoom to more than 100% of the actual image size. It is not visible at every border or every level but is very visible when there is text/symbols across 2 tiles.
I downloaded the tiles, stitched them manually and the image was fine, so the tile source is not the problem.
I think this is caused because each tile is rendered in a separate canvas and the browser resizes/optimizes the tile images (using bicubic or something) separately. This may not be fixable without compromising performance, but it is worth a try.

The example below is zoomed to ~800% for demonstration but the issue is apparent even at 100%
Example:
![viewer-bug](https://cloud.githubusercontent.com/assets/5219205/10790017/91d032fa-7d89-11e5-9454-cb21abeb045f.jpg)

**Comment by avandecreme** [comment]
Do you have any tiles overlap?
Can you try to increase the overlap?

**Comment by MindFreeze** [comment]
There is no overlap. The placement seems to be correct.

**Comment by avandecreme** [comment]
>There is no overlap.

I suspect that adding overlap would fix the issue. Can you try that?

>I can upload the tiled image that caused this, but it is a JPEG2000 . Don't know if you can work with that.

Not really but if it is a dzi, I can convert it to add overlap.

**Comment by MindFreeze** [comment]
The thing is, I can only work with jp2. And can't see an option to add overlap for that. Is it only possible for dzi?

**Comment by avandecreme** [comment]
Ho ok, I actually don't know how IIIF works.
Is it possible to tell IIIF to provide overlap?
Maybe @azaroth42 or @jpstroop can help here?

**Comment by MindFreeze** [comment]
My server is actually http://iipimage.sourceforge.net/ So I can ditch IIIF and create a custom protocol for it (didn't see one for IIP). Then I may be able to add overlap somehow. But I don't know when I would have enough time to try it.

**Comment by iangilman** [comment]
Tile overlap may help, or it might not. The artifact you highlighted in that image seems pretty fundamental to scaling the tiles separately; since they don't know that they will be blended with their neighbor, they have a sharp edge. 

Back in the day, when the Silverlight Seadragon used DirectX, we used tile overlap and feathered the edges to avoid that seam. With OpenSeadragon we support tile overlap, but I don't think we feather edges even when it's present; I believe that would indeed have an impact on our performance with Canvas. Another approach, I suppose, would be to do the compositing at 1:1 and then scale up as needed. This would probably also have a performance impact. 

That said, the effect I'm describing should only be visible at greater than 100%; it should be completely invisible at 100%, so if you're still seeing it at 100%, I would look for another explanation. I think it's definitely worth converting your image to other formats (including experimenting with tile overlap), to at least eliminate that variable. I've certainly worked with many DZI images that don't seem to have this issue.

**Comment by MindFreeze** [comment]
I will try to tinker with the zoom a bit more tomorrow. I may not have been at exactly 100% because to do it, I have to use `viewer.viewport.imageToViewportZoom(1)` and that might loose accuracy do to floating points. I also think it is visible at 100% but that might be because I know it is there. 

In any case the issue is there at >100% and the ability to zoom like that is 1 of the main reasons we started using OpenSeadragon at our company. 

Maybe we add a less general fix like use feathering only if zoom is >100% and we know there is overlap. Or feathering, blending, compositing at 1:1 or whatever could be added as an option so it doesn't affect performance unless someone needs it. And yes, I realise how much work these are but this seems like a serious flaw.

Btw I also see the problem here http://picturae.github.io/openseadragonselection/ and this is a dzi from the example images. Though that may be the other way around because there is overlap=2 . The tileSource is:
```
       {
            Image: {
                xmlns: 'http://schemas.microsoft.com/deepzoom/2008',
                Url: 'http://openseadragon.github.io/example-images/duomo/duomo_files/',
                Format: 'jpg',
                Overlap: '2',
                TileSize: '256',
                Size: {
                    Width:  '13920',
                    Height: '10200'
                }
            }
        }
```

**Comment by azaroth42** [comment]
> the IIIF tilesource could add overlap by changing the requested dimensions and the server will stitch tiles where needed.

This is correct. You could request overlapping tiles by manipulating the region parameter. Unless it can be demonstrated that this is actually the problem, I wouldn't recommend doing it however, as the performance will suffer due to lack of caching.

I don't think this is IIIF related -- we don't see this with other images -- but demonstrations and test cases are always welcome!

**Comment by avandecreme** [comment]
Yeah definitly not IIIF related, I was wondering if the lack of overlap might be the issue.

@MindFreeze I don't see the problem on the OpenSeadragon demo image.
We can see a line here:
![line](https://cloud.githubusercontent.com/assets/465389/10802529/c7606bba-7d93-11e5-8a00-f82e8d5e1497.png)

But it is also present in the tiles:
![tile](http://openseadragon.github.io/example-images/duomo/duomo_files/14/18_13.jpg)

Are you seeing something else?

**Comment by MindFreeze** [comment]
> I don't see the problem on the OpenSeadragon demo image.
As I suspected that turned out to be a bad overlap value. It is fixed now. But the jp2 problem is unrelated.

> Yeah definitly not IIIF related, I was wondering if the lack of overlap might be the issue.
It is not a problem with the protocol but with the way OpenSeadragon renders the tiles.

When there is even 1px overlap the canvas zooms correctly and there is no tearing. I just tried it with the demo page. So if I add overlap through customizing IIIF it would fix the issue in my case but would impact performance on the server side.

**Comment by MindFreeze** [comment]
Edges are not visible at exactly 100% zoom but the zoom levels are quite dynamic so that does not help that much. And edges are visible at 95-99% and >100% .

The canvas seems to regard the image as whole if the tiles overlap. So can't we instruct/fool the canvas to do that even if they don't overlap?

**Comment by MindFreeze** [comment]
I tried drawing the tiles directly on the big canvas without using separate canvases by skipping `getRenderedContext()` and doing this in `tile.js:292` :
```
        context.drawImage(
            this.cacheImageRecord._image,
            0,
            0,
            this.cacheImageRecord._image.width,
            this.cacheImageRecord._image.height,
            position.x * $.pixelDensityRatio,
            position.y * $.pixelDensityRatio,
            size.x * $.pixelDensityRatio,
            size.y * $.pixelDensityRatio
        );
```

Didn't help. Still see edges.

**Comment by iangilman** [comment]
Ok, I don't think it has anything to do with your tile format or overlap; it's just a matter of zooming in far enough. Here is the Duomo DZI zoomed in a lot: 

<img width="807" alt="screen shot 2015-10-29 at 9 46 15 am" src="https://cloud.githubusercontent.com/assets/338989/10825388/0ba04d16-7e22-11e5-8d41-300bdbcaa472.png">

I think in general people don't set their viewers up to allow zooming much beyond 100% (the default maxZoomPixelRatio is 1.1, after all). That's why people haven't really run into this before. 

So yes, I think the solution is to add some special casing for the over 100% scenario. I think the best option is going to be compositing the tiles at 100% and then scaling them up together. While edge feathering would be nice, it only applies if you have overlap. 

We already have a second canvas available for doing this sort of compositing; we currently use it for times when an image has less than 100% opacity. It's called the sketch canvas, and you can see it in use here (follow the `useSketch` variable): 

https://github.com/openseadragon/openseadragon/blob/master/src/tiledimage.js#L1300-L1360

@MindFreeze Are you up for making a patch for this?

**Comment by MindFreeze** [comment]
Sure. I will give it a stab when I have the time.

**Comment by iangilman** [comment]
Excellent. Let me know when you have questions!

**Comment by iangilman** [comment]
By the way, looking into this, it looks like the Duomo image has some other problems, like maybe I didn't do a very good job of converting it; if you just sit and watch it load, the tiles kind of seem to bounce around a bit. Our "Highsmith" image is much more solid. Likewise if I zoom in on Highsmith, while I can see the seams, they are definitely less noticeable than on Duomo.

**Comment by iangilman** [comment]
Fixed by #764
